File: code/0098.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def check(root):
            has_right = root.right is not None
            has_left = root.left is not None
            if has_right:
                right_bst, right_max, right_min = check(root.right)
            else:
                right_bst, right_max = True, root.val
            if has_left:
                left_bst, left_max, left_min = check(root.left)
            else:
                left_bst, left_min = True, root.val
            root_bst = ((not has_left) or left_max < root.val) and \
                        ((not has_right) or root.val < right_min)
            return left_bst and root_bst and right_bst, right_max, left_min
        [out_res, _, _] = check(root)
        return out_res

File: code/0098.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<int> vec;
        check(root,vec);
        for(int i = 1; i < vec.size() ; ++i)
            if(vec[i] <= vec[i-1])
                return false;
        return true;
    }
    void check(TreeNode* root, vector<int>& vec){
        if(root->left != nullptr) check(root->left,vec);
        vec.push_back(root->val);
        if(root->right != nullptr) check(root->right,vec);
    }
};